core/modules/node/node.pages.inc

```diff
@@ -444,7 +444,21 @@ function node_form_submit($form, &$form_state) {
   if ($node->nid) {
     $form_state['values']['nid'] = $node->nid;
     $form_state['nid'] = $node->nid;
-    $form_state['redirect'] = node_access('view', $node) ? 'node/' . $node->nid : '<front>';
+
+    // Determine whether to redirect user to full page display.
+    $form_state['redirect'] = '<front>';
+    if (node_access('view', $node)) {
+      $form_state['redirect'] = 'node/' . $node->nid;
+
+      // If node type set to prevent direct viewing of full page URL
+      // stay on edit page if user is not permitted.
+      $hidden_path = config_get('node.type.' . $node->type, 'settings.hidden_path');
+      $bypass_hidden_path = user_access('view hidden path display');
+      if ($hidden_path && !$bypass_hidden_path) {
+        $form_state['redirect'] = 'node/' . $node->nid . '/edit';
+        backdrop_set_message(t('You do not have permission to view content where the path is hidden.'), 'warning');
```

Comment:
I don't think we should show this message. As there's nothing they can do to correct it, and there's nothing actually wrong here.